debezium-server/debezium-server-ybexporter/src/main/java/io/debezium/server/ybexporter/StreamingWriterJson.java

```diff
@@ -5,87 +5,141 @@
+ * segment that was written to, and continues writing from that segment.
+ */
 public class StreamingWriterJson implements RecordWriter {
-    private static final Logger LOGGER = LoggerFactory.getLogger(TableSnapshotWriterCSV.class);
-    private static final String QUEUE_FILE_NAME = "queue.json";
+    private static final Logger LOGGER = LoggerFactory.getLogger(StreamingWriterJson.class);
+    private static final String QUEUE_FILE_NAME = "queue";
+    private static final String QUEUE_FILE_EXTENSION = "ndjson";
+    private static final String QUEUE_FILE_DIR = "cdc";
+    private long queueSegmentMaxBytes = 1000 * 1000 * 1000; // default 1 GB
     private String dataDir;
-    private BufferedWriter writer;
-    private ObjectWriter ow;
-    private FileOutputStream fos;
-    private FileDescriptor fd;
+    private QueueSegment currentQueueSegment;
+    private long currentQueueSegmentIndex = 0;
 
-    public StreamingWriterJson(String datadirStr) {
-        dataDir = datadirStr;
 
-        var fileName = String.format("%s/%s", dataDir, QUEUE_FILE_NAME);
-        try {
-            fos = new FileOutputStream(fileName, true);
-            fd = fos.getFD();
-            var f = new FileWriter(fd);
-            writer = new BufferedWriter(f);
+    public StreamingWriterJson(String datadirStr, Long queueSegmentMaxBytes) {
+        dataDir = datadirStr;
+        if (queueSegmentMaxBytes != null){
+            this.queueSegmentMaxBytes = queueSegmentMaxBytes;
+        }
 
+        // mkdir cdc
+        File queueDir = new File(String.format("%s/%s", dataDir, QUEUE_FILE_DIR));
+        if (!queueDir.exists()){
+            boolean dirCreated = queueDir.mkdir();
+            if (!dirCreated){
+                throw new RuntimeException("failed to create dir for cdc");
+            }
         }
-        catch (IOException e) {
-            throw new RuntimeException(e);
+        recoverStateFromDisk();
+        if (currentQueueSegment == null){
+            createNewQueueSegment();
         }
-        ow = new ObjectMapper().writer().withDefaultPrettyPrinter();
     }
 
-    @Override
-    public void writeRecord(Record r) {
+    private void createNewQueueSegment(){
+        currentQueueSegment = new QueueSegment(getFilePathWithIndex(currentQueueSegmentIndex));
+    }
+
+    private void recoverStateFromDisk(){
+        recoverLatestQueueSegment();
+    }
+
+    /**
+     * This function reads the /cdc dir for all the queue segment files.
+     * Then it retrieves the index number from the paths, and then finds
+     * the max index - which is the latest queue segment that was written to.
+     * If no files are found, we just return.
+     */
+    private void recoverLatestQueueSegment(){
+        // read dir to find all queue files
+        Path queueDirPath = Path.of(dataDir, QUEUE_FILE_DIR);
+        String searchGlob = String.format("%s.[0-9]*.%s", QUEUE_FILE_NAME, QUEUE_FILE_EXTENSION);
```

Comment:
this might match with something like `queue.0.ndjson.processed`
make sure the check end at `ndjson`, for that i think `$` is used.
`%s.[0-9]+.%s$`

Reply:
hah, glob doesn't seem to have anything like that to match end of string. Current glob seems to work and will not include something like `queue.0.ndjson.processed`

```
» ls              
queue.0.ndjson           queue.2.ndjson
queue.1.ndjson           queue.2.ndjson.processed
----------------------------------------------------------------------------
» ls queue.[0-9]*.ndjson
queue.0.ndjson queue.1.ndjson queue.2.ndjson
----------------------------------------------------------------------------
» ls queue.[0-9]*.ndjson$
zsh: no matches found: queue.[0-9]*.ndjson$
```

Reply:
oh ok good, actually in case i made a regex and compared it with all the filenames fetched from the directory and was causing this issue